src/main/java/io/cryostat/rules/PeriodicArchiver.java

```diff
@@ -91,6 +97,34 @@ public void run() {
         logger.trace("PeriodicArchiver for {} running", rule.getRecordingName());
 
         try {
+            // If there are no previous recordings, either this is the first time this rule is being
+            // archived
+            // or the Cryostat instance was restarted. Since it could be the latter, populate the
+            // array
+            // with any previously archived recordings for this rule.
+            if (previousRecordings.isEmpty()) {
+                JsonArray archivedRecordings = getArchivedRecordings().get();
+                Iterator<Object> it = archivedRecordings.iterator();
+
+                URI serviceUri = serviceRef.getServiceUri();
+                Pattern recordingFilenamePattern =
+                        Pattern.compile(
+                                String.format(
+                                        "([A-Za-z\\d-]*)_(%s)_([\\d]*T[\\d]*Z)(\\.[\\d]+)?",
```

Comment:
Instead of building the regex pattern with the serviceUri inserted into the middle of it, this can be a constant pattern with a capture group where the serviceUri is expected. Then the captured value of that group can be pulled out to compare to the expected serviceUri, or hashed, or whatever else is needed.

Reply:
It looks like this pattern probably originally came from: https://github.com/cryostatio/cryostat/blob/af47f3f5b08c673cb1df7cae8c739db5b2f597b8/src/main/java/io/cryostat/net/web/http/api/v1/RecordingsPostHandler.java#L76

The matcher and capture group stuff can also be seen in use with that pattern here:

https://github.com/cryostatio/cryostat/blob/af47f3f5b08c673cb1df7cae8c739db5b2f597b8/src/main/java/io/cryostat/net/web/http/api/v1/RecordingsPostHandler.java#L164

https://github.com/cryostatio/cryostat/blob/af47f3f5b08c673cb1df7cae8c739db5b2f597b8/src/main/java/io/cryostat/net/web/http/api/v1/RecordingsPostHandler.java#L169

Reply:
Just to clarify, the regex pattern has the recording name (in this case, `auto_$(Rule.name)`) inserted into the middle, not the serviceUri. The serviceUri hash is compared directly inside the loop, and not with the matcher, since the serviceUri is not part of the overall filename but just the parent directory.

Capturing seems more robust, I'll take a look at those lines, thanks.

Reply:
Ah right, makes sense. In any case, I think the capture group way is more conventional and so probably easier to read/maintain in the future. As an added bonus it's only one pattern that needs to be compiled at application startup, rather than a new pattern to be compiled on every request, so there should be a (very slight) performance benefit too.

Reply:
Sounds good, I'll implement the capture group regardless.